apps/vscode/command_client.py

```diff
@@ -1,126 +1,235 @@
+
+    Args:
+        path (Path): The path of the file to write
+        body (Any): The object to convert to json and write
+    """
+    with path.open("x") as out_file:
+        out_file.write(json.dumps(body))
+
+
+@dataclass
+class Request:
+    command_id: str
+    args: List[Any]
+    wait_for_finish: bool
+    return_command_output: bool
+    uuid: str
+
+    def to_dict(self):
+        return {
+            "commandId": self.command_id,
+            "args": self.args,
+            "waitForFinish": self.wait_for_finish,
+            "returnCommandOutput": self.return_command_output,
+            "uuid": self.uuid,
+        }
+
+
+def write_request(request: Request, path: Path):
+    """Converts the given request to json and writes it to the file, failing if
+    the file already exists unless it is stale in which case it replaces it
+
+    Args:
+        request (Request): The request to serialize
+        path (Path): The path to write to
+
+    Raises:
+        Exception: If another process has an active request file
+    """
+    try:
+        write_json_exclusive(path, request.to_dict())
+    except FileExistsError:
+        stats = path.stat()
+
+        modified_time_ms = stats.st_mtime_ns / 1e6
+        current_time_ms = time.time() * 1e3
+
+        if abs(modified_time_ms - current_time_ms) < STALE_TIMEOUT_MS:
+            raise Exception("Another process has an active request file")
+        else:
+            print("Removing stale request file")
+            path.unlink()
+            write_json_exclusive(path, request.to_dict())
+
+
 def run_vscode_command(
-    command: str,
+    command_id: str,
     *args: str,
     wait_for_finish: bool = False,
-    expect_response: bool = False,
+    return_command_output: bool = False,
     decode_json_arguments: bool = False,
 ):
-    """Execute command via vscode command server."""
+    """Runs a VSCode command, using command server if available
+
+    Args:
+        command (str): The ID of the VSCode command to run
+        wait_for_finish (bool, optional): Whether to wait for the command to finish before returning. Defaults to False.
+        return_command_output (bool, optional): Whether to return the output of the command. Defaults to False.
+        decode_json_arguments (bool, optional): Whether to decode JSON arguments. Defaults to False.
+
+    Raises:
+        Exception: If there is an issue with the file-based communication
+
+    Returns:
+        Object: The response from the command, if requested.
+    """
     # NB: This is a hack to work around the fact that talon doesn't support
     # variable argument lists
-    args = list(
-        filter(
-            lambda x: x is not NotSet,
-            args,
-        )
-    )
+    args = [x for x in args if x is not NotSet]
+
     if decode_json_arguments:
         args = [json.loads(arg) for arg in args]
 
-    port_file_path = Path(gettempdir()) / "vscode-port"
+    username = getpass.getuser()
```

Comment:
I don't think the UID approach will be portable to Windows, so would make this code more complicated, and I'm not sure I see the downside of using environment variables in this situation.  We're not using this username for access control, just a convenient way to make the communication dir user-specific

Reply:
My concern is about quality, not access control. I've personally had users mess up a lot of stuff in their environment you wouldn't expect.

On Windows, the temp dir is user-specific anyway, so we don't need to suffix there. I also found this about `userInfo().username` on the node side:

> The only problem is, in Windows 10, it returns the first name of the owner account that has been used (just a heads up).

My vote would be to use getuid if available, otherwise don't suffix the path at all, roughly like this:

```
suffix = ""
if hasattr(os, "getuid"):
    suffix = "-{os.getuid()}"
communication_dir_path = Path(gettempdir()) / f"vscode-command-server{suffix}"
```

On the node side, you can call `userInfo()` and do something like this:
```
if (info.uid >= 0) path += `-${info.uid}`
```
or whatever, because
> On Windows, the uid and gid fields are -1, and shell is null.

Reply:
this is great thanks @lunixbochs